File: scraper/scrape_menus.py
"""
Scrapes full menu text for cafes from two sources (text only, no PDFs/images):
  1. Google Maps — "Menu" tab on the place page
  2. Cafe website — finds menu page, extracts body text

Only saves menus that look complete (enough items / length).
Saves progress to data/menu_progress.json (resumable).
Writes results to public/cafes.json when done.
"""

import asyncio
import io
import json
import re
import sys
from pathlib import Path

sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8", errors="replace")

from playwright.async_api import async_playwright

ROOT        = Path(__file__).parent.parent
CAFES_FILE  = ROOT / "public" / "cafes.json"
PROGRESS    = ROOT / "data" / "menu_progress2.json"
RESULTS     = ROOT / "data" / "menu_results2.json"

# ── Helpers ────────────────────────────────────────────────────────────────────

FOOD_WORDS = [
    "coffee", "espresso", "latte", "cappuccino", "flat white", "long black",
    "oat milk", "almond milk", "soy milk", "matcha", "chai", "hot chocolate",
    "cold brew", "filter", "pour over", "batch brew",
    "toast", "eggs", "avocado", "smashed avo", "benedict", "poached",
    "scrambled", "fried", "omelette", "frittata",
    "smoothie", "juice", "açaí", "acai",
    "cake", "muffin", "croissant", "danish", "pastry", "scone",
    "sandwich", "wrap", "salad", "bowl", "granola", "bircher",
    "sourdough", "rye", "focaccia", "brioche",
    "lunch", "brunch", "breakfast", "snack",
]

PRICE_RE = re.compile(r"\$\s*\d+(?:\.\d{2})?")
MENU_LINK_RE = re.compile(
    r"\b(menu|food|drink|eat|brunch|breakfast|what we serve|order)\b", re.I
)


def food_hit_count(text):
    lower = text.lower()
    return sum(1 for w in FOOD_WORDS if w in lower)


def is_full_menu(text):
    """Return True only if text looks like a real, full menu."""
    if not text or len(text) < 400:
        return False
    hits = food_hit_count(text)
    has_prices = bool(PRICE_RE.search(text))
    lines = [l.strip() for l in text.splitlines() if l.strip()]
    # Need either many food words OR prices + some food words
    if has_prices and hits >= 4:
        return True
    if hits >= 8 and len(lines) >= 10:
        return True
    return False


def clean_text(raw):
    raw = re.sub(r"[ \t]+", " ", raw)
    raw = re.sub(r"\n{3,}", "\n\n", raw)
    return raw.strip()[:5000]


# ── Google Maps menu scrape ────────────────────────────────────────────────────

async def scrape_gmaps_menu(page, cafe):
    url = cafe.get("googleMapsUrl") or (
        f"https://www.google.com/maps/search/"
        f"{cafe['name'].replace(' ', '+')}+{cafe['suburb'].replace(' ', '+')}+Melbourne"
    )
    try:
        await page.goto(url, wait_until="domcontentloaded", timeout=20000)
        await page.wait_for_timeout(2500)
    except Exception:
        return None

    # If on search results page, click first result
    if "@" not in page.url:
        try:
            first = page.locator('a[href*="/maps/place/"]').first
            await first.click(timeout=4000)
            await page.wait_for_timeout(2000)
        except Exception:
            return None

    # Look for a "Menu" tab/button
    for label in ["Menu", "See menu", "Full menu"]:
        try:
            btn = page.locator(f'button:has-text("{label}"), a:has-text("{label}")').first
            await btn.click(timeout=3000)
            await page.wait_for_timeout(2000)
            break
        except Exception:
            continue

    # Extract inner text of the place panel
    try:
        # The main panel in Google Maps is the left sidebar
        panel = page.locator('[role="main"]').first
        text = await panel.inner_text(timeout=4000)
        text = clean_text(text)
        if is_full_menu(text):
            return text
    except Exception:
        pass

    return None


# ── Website menu scrape ───────────────────────────────────────────────────────

MENU_PATH_KEYWORDS = [
    "/menu", "/menus", "/food", "/food-menu", "/drink-menu", "/drinks",
    "/our-menu", "/cafe-menu", "/breakfast", "/brunch", "/coffee",
    "/what-we-serve", "/eat", "/food-and-drink", "/order",
]


async def scrape_website_menu(page, website):
    base = website.rstrip("/")
    try:
        await page.goto(base, wait_until="domcontentloaded", timeout=15000)
        await page.wait_for_timeout(1500)
    except Exception:
        return None

    # Collect menu-like links from homepage
    try:
        links = await page.evaluate("""(function(base) {
            return [...new Set(
                Array.from(document.querySelectorAll('a[href]'))
                    .map(a => {
                        try { return new URL(a.getAttribute('href'), base).href; } catch { return ''; }
                    })
                    .filter(h => {
                        if (!h || !h.startsWith(base)) return false;
                        const p = new URL(h).pathname.toLowerCase();
                        return /menu|food|drink|eat|brunch|breakfast|order/.test(p)
                            && !p.endsWith('.pdf') && !p.endsWith('.jpg')
                            && !p.endsWith('.png') && !p.endsWith('.webp');
                    })
            )].slice(0, 6);
        })(arguments[0])""", base)
    except Exception:
        links = []

    # Add common path guesses
    guesses = [base + p for p in MENU_PATH_KEYWORDS
               if not any(p in l for l in links)]
    to_try = (links + guesses)[:8]

    # Try homepage first
    for url in [base] + to_try:
        try:
            if url != base:
                await page.goto(url, wait_until="domcontentloaded", timeout=12000)
                await page.wait_for_timeout(1000)

            # Skip if PDF or binary (check content type via URL)
            if re.search(r"\.(pdf|jpg|jpeg|png|gif|webp|svg)$", url, re.I):
                continue

            text = await page.evaluate("""(function() {
                const remove = ['script','style','noscript','header','footer',
                                'nav','iframe','form','aside'];
                remove.forEach(t => document.querySelectorAll(t)
                    .forEach(el => el.remove()));
                // Prefer menu/food sections
                for (const sel of ['[class*="menu"]','[id*="menu"]',
                                   '[class*="food"]','main','article','#content']) {
                    const el = document.querySelector(sel);
                    if (el && el.innerText.trim().length > 300) return el.innerText;
                }
                return document.body.innerText;
            })()""")

            text = clean_text(text or "")
            if is_full_menu(text):
                return text
        except Exception:
            continue

    return None


# ── Main ──────────────────────────────────────────────────────────────────────

async def main():
    cafes  = json.loads(CAFES_FILE.read_text(encoding="utf-8"))
    prog   = json.loads(PROGRESS.read_text(encoding="utf-8")) if PROGRESS.exists() else {}
    done   = set(prog.keys())
    todo   = [c for c in cafes if c["id"] not in done]

    print(f"Total: {len(cafes)} | Done: {len(done)} | To do: {len(todo)}")

    found = sum(1 for v in prog.values() if v)
    print(f"Menus found so far: {found}")

    if not todo:
        print("All done — writing to cafes.json")
    else:
        async with async_playwright() as p:
            browser = await p.chromium.launch(headless=True)
            context = await browser.new_context(
                user_agent=(
                    "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
                    "AppleWebKit/537.36 (KHTML, like Gecko) "
                    "Chrome/124.0.0.0 Safari/537.36"
                ),
                viewport={"width": 1280, "height": 900},
                locale="en-AU",
            )
            page = await context.new_page()

            for i, cafe in enumerate(todo):
                print(f"\n{len(done)+1}/{len(cafes)} | {cafe['name']} — {cafe['suburb']}")

                menu = None

                # 1. Try Google Maps first
                menu = await scrape_gmaps_menu(page, cafe)
                if menu:
                    print(f"  [gmaps] {len(menu)} chars")

                # 2. Try website if no menu from Maps
                if not menu and cafe.get("website"):
                    menu = await scrape_website_menu(page, cafe["website"])
                    if menu:
                        print(f"  [web]   {len(menu)} chars")

                if not menu:
                    print(f"  no menu found")

                prog[cafe["id"]] = menu
                done.add(cafe["id"])
                if menu:
                    found += 1

                if len(done) % 20 == 0:
                    PROGRESS.write_text(
                        json.dumps(prog, indent=2, ensure_ascii=False), encoding="utf-8"
                    )
                    print(f"  [saved — {found} menus so far]")

                await asyncio.sleep(1.5)

            await browser.close()

        PROGRESS.write_text(json.dumps(prog, indent=2, ensure_ascii=False), encoding="utf-8")

    # Write results file (found only)
    results = {k: v for k, v in prog.items() if v}
    RESULTS.parent.mkdir(exist_ok=True)
    RESULTS.write_text(json.dumps(results, indent=2, ensure_ascii=False), encoding="utf-8")
    print(f"\nMenu results: {len(results)} cafes")

    # Apply to cafes.json
    cafe_map = {c["id"]: c for c in cafes}
    patched = 0
    for cid, text in results.items():
        if cid in cafe_map and not cafe_map[cid].get("menuText"):
            cafe_map[cid]["menuText"] = text
            patched += 1

    CAFES_FILE.write_text(
        json.dumps(cafes, indent=2, ensure_ascii=False), encoding="utf-8"
    )
    print(f"Written menuText to {patched} cafes in cafes.json")


if __name__ == "__main__":
    asyncio.run(main())

File: scripts/scrape_menus.js
/**
 * scrape_menus.js
 * Scrapes menu text from cafe websites.
 * Saves progress to data/menu_progress.json (resumable).
 * Run: node scripts/scrape_menus.js
 */

import fs   from 'fs';
import path from 'path';
import https from 'https';
import http  from 'http';
import { fileURLToPath } from 'url';
import { load } from 'cheerio';

const __dirname = path.dirname(fileURLToPath(import.meta.url));

const CAFES_FILE    = path.join(__dirname, '../public/cafes.json');
const PROGRESS_FILE = path.join(__dirname, '../data/menu_progress.json');
const OUT_FILE      = path.join(__dirname, '../data/menu_results.json');

const TIMEOUT_MS  = 8000;
const CONCURRENCY = 5;
const DELAY_MS    = 300;

const MENU_PATHS = [
  '/menu', '/menus', '/food', '/food-menu', '/drink-menu', '/drinks',
  '/our-menu', '/cafe-menu', '/breakfast', '/brunch', '/coffee',
  '/what-we-serve', '/eat', '/food-and-drink',
];

const MENU_LINK_RE = /\b(menu|food|drinks?|eat|brunch|breakfast|what we (serve|offer))\b/i;

// ── HTTP fetch ────────────────────────────────────────────────────────────────

function fetchUrl(url, timeoutMs = TIMEOUT_MS, redirectCount = 0) {
  return new Promise((resolve, reject) => {
    if (redirectCount > 3) return resolve(null);
    let parsed;
    try { parsed = new URL(url); } catch { return resolve(null); }
    const lib = parsed.protocol === 'https:' ? https : http;
    const req = lib.get({
      hostname: parsed.hostname,
      path:     parsed.pathname + parsed.search,
      headers:  {
        'User-Agent': 'Mozilla/5.0 (compatible; MelbCafeFinder/1.0)',
        Accept:       'text/html,application/pdf',
      },
      timeout: timeoutMs,
    }, (res) => {
      if ([301, 302, 303, 307, 308].includes(res.statusCode) && res.headers.location) {
        try {
          const next = new URL(res.headers.location, url).href;
          return fetchUrl(next, timeoutMs, redirectCount + 1).then(resolve).catch(reject);
        } catch { return resolve(null); }
      }
      if (res.statusCode !== 200) return resolve(null);
      const ct = res.headers['content-type'] || '';
      const chunks = [];
      res.on('data', (c) => chunks.push(c));
      res.on('end',  () => resolve({ body: Buffer.concat(chunks), contentType: ct }));
      res.on('error', reject);
    });
    req.on('error',   reject);
    req.on('timeout', () => { req.destroy(); reject(new Error('timeout')); });
  });
}

// ── HTML → clean text ─────────────────────────────────────────────────────────

function extractText(html) {
  const $ = load(html);
  $('script, style, noscript, header, footer, nav, iframe, form').remove();
  $('[class*="cookie"],[class*="popup"],[class*="modal"],[class*="banner"],[class*="newsletter"]').remove();

  const menuSelectors = [
    '[class*="menu"]', '[id*="menu"]',
    '[class*="food"]',  '[id*="food"]',
    '[class*="drink"]', '[id*="drink"]',
    'main', 'article', '.content', '#content',
  ];

  let text = '';
  for (const sel of menuSelectors) {
    const el = $(sel).first();
    if (el.length) {
      text = el.text();
      if (text.trim().length > 200) break;
    }
  }
  if (!text || text.trim().length < 100) text = $('body').text();

  return text.replace(/[ \t]+/g, ' ').replace(/\n{3,}/g, '\n\n').trim().slice(0, 3000);
}

// ── Find links on homepage ────────────────────────────────────────────────────

function findMenuLinks(html, baseUrl) {
  const $ = load(html);
  const links = [];
  $('a[href]').each((_, el) => {
    const href = $(el).attr('href');
    const text = $(el).text();
    if (!href || /^(#|mailto:|tel:)/.test(href)) return;
    if (MENU_LINK_RE.test(text) || MENU_LINK_RE.test(href)) {
      try {
        const abs = new URL(href, baseUrl).href;
        if (new URL(abs).hostname === new URL(baseUrl).hostname) links.push(abs);
      } catch {}
    }
  });
  return [...new Set(links)].slice(0, 3);
}

function findPdfLinks(html, baseUrl) {
  const $ = load(html);
  const links = [];
  $('a[href]').each((_, el) => {
    const href = $(el).attr('href') || '';
    if (!href.toLowerCase().endsWith('.pdf')) return;
    try { links.push(new URL(href, baseUrl).href); } catch {}
  });
  return links.slice(0, 2);
}

// ── PDF text extraction ───────────────────────────────────────────────────────

async function extractPdf(buf) {
  try {
    const { default: pdfParse } = await import('pdf-parse/lib/pdf-parse.js');
    const data = await pdfParse(buf);
    return data.text.replace(/\s+/g, ' ').trim().slice(0, 3000);
  } catch {
    return null;
  }
}

// ── Menu heuristic ────────────────────────────────────────────────────────────

function isLikelyMenu(text) {
  if (!text || text.length < 150) return false;
  const lower = text.toLowerCase();
  const hits = ['coffee', 'espresso', 'latte', 'cappuccino', 'flat white',
    'toast', 'eggs', 'avocado', 'smoothie', 'juice', 'cake', 'muffin',
    'sandwich', 'matcha', 'chai', 'oat milk', 'almond', 'soy',
    'croissant', 'pastry', 'breakfast', 'brunch', 'lunch', 'bowl',
    'wrap', 'salad', 'granola', 'acai', 'benedict', 'poached',
  ].filter(w => lower.includes(w)).length;
  return hits >= 3;
}

// ── Scrape one cafe ───────────────────────────────────────────────────────────

async function scrapeMenu(website) {
  const base = website.replace(/\/$/, '');

  let homeRes;
  try { homeRes = await fetchUrl(base); } catch { return null; }
  if (!homeRes) return null;

  const homeHtml = homeRes.body.toString('utf8');

  // Check for PDF links on homepage
  for (const pdfUrl of findPdfLinks(homeHtml, base)) {
    try {
      const res = await fetchUrl(pdfUrl);
      if (res?.contentType.includes('pdf')) {
        const text = await extractPdf(res.body);
        if (text && text.length > 100) return text;
      }
    } catch {}
  }

  // Try menu links found on homepage + common paths
  const menuLinks = findMenuLinks(homeHtml, base);
  const toTry = [...new Set([...menuLinks, ...MENU_PATHS.map(p => base + p)])].slice(0, 10);

  for (const url of toTry) {
    try {
      const res = await fetchUrl(url);
      if (!res) continue;
      if (res.contentType.includes('pdf')) {
        const text = await extractPdf(res.body);
        if (text && text.length > 150) return text;
      } else if (res.contentType.includes('html')) {
        const text = extractText(res.body.toString('utf8'));
        if (isLikelyMenu(text)) return text;
      }
    } catch {}
    await delay(80);
  }

  // Fall back: homepage itself
  const homeText = extractText(homeHtml);
  return isLikelyMenu(homeText) ? homeText : null;
}

function delay(ms) { return new Promise(r => setTimeout(r, ms)); }

// ── Main ──────────────────────────────────────────────────────────────────────

async function main() {
  const cafes    = JSON.parse(fs.readFileSync(CAFES_FILE, 'utf8'));
  const progress = fs.existsSync(PROGRESS_FILE)
    ? JSON.parse(fs.readFileSync(PROGRESS_FILE, 'utf8'))
    : {};

  const withSite = cafes.filter(c => c.website);
  const todo     = withSite.filter(c => !(c.id in progress));

  console.log(`Total cafes:    ${cafes.length}`);
  console.log(`With website:   ${withSite.length}`);
  console.log(`Already done:   ${Object.keys(progress).length}`);
  console.log(`To scrape:      ${todo.length}\n`);

  let done = 0, found = 0;
  const save = () => fs.writeFileSync(PROGRESS_FILE, JSON.stringify(progress, null, 2));

  for (let i = 0; i < todo.length; i += CONCURRENCY) {
    const batch = todo.slice(i, i + CONCURRENCY);
    await Promise.all(batch.map(async (cafe) => {
      try {
        const text = await scrapeMenu(cafe.website);
        progress[cafe.id] = text || null;
        if (text) found++;
      } catch {
        progress[cafe.id] = null;
      }
      done++;
      process.stdout.write(`\r${done}/${todo.length} — ${found} menus found`);
    }));
    save();
    await delay(DELAY_MS);
  }

  save();
  console.log(`\n\nDone! ${found} menus found out of ${withSite.length} cafes with websites.`);

  // Write found-only results
  const results = Object.fromEntries(
    Object.entries(progress).filter(([, v]) => v)
  );
  fs.writeFileSync(OUT_FILE, JSON.stringify(results, null, 2));
  console.log(`Saved to data/menu_results.json`);
}

main().catch(console.error);
